sudoku.rb

```diff
@@ -0,0 +1,165 @@
+# Sudoku Rules:
+#   Each cell should contain a digit 1-9 such that:
+#     * horizontal row contains each digit exactly once
+#       [_ _ _  _ _ _  _ _ _]
+#       [_ _ _  _ _ _  _ _ _]
+#       [_ _ _  _ _ _  _ _ _]
+#
+#       [_ _ _  _ _ _  _ _ _]
+#       [1 2 3  4 5 6  7 8 9]
+#       [_ _ _  _ _ _  _ _ _]
+#
+#       [_ _ _  _ _ _  _ _ _]
+#       [_ _ _  _ _ _  _ _ _]
+#       [_ _ _  _ _ _  _ _ _]
+#     * vertical column contains each digit exactly once
+#       [_ _ _  _ 1 _  _ _ _]
+#       [_ _ _  _ 2 _  _ _ _]
+#       [_ _ _  _ 3 _  _ _ _]
+#
+#       [_ _ _  _ 4 _  _ _ _]
+#       [_ _ _  _ 5 _  _ _ _]
+#       [_ _ _  _ 6 _  _ _ _]
+#
+#       [_ _ _  _ 7 _  _ _ _]
+#       [_ _ _  _ 8 _  _ _ _]
+#       [_ _ _  _ 9 _  _ _ _]
+#     * each 3x3 box contains each digit exactly once
+#       [_ _ _  _ _ _  _ _ _]
+#       [_ _ _  _ _ _  _ _ _]
+#       [_ _ _  _ _ _  _ _ _]
+#
+#       [_ _ _  1 2 3  _ _ _]
+#       [_ _ _  4 5 6  _ _ _]
+#       [_ _ _  7 8 9  _ _ _]
+#
+#       [_ _ _  _ _ _  _ _ _]
+#       [_ _ _  _ _ _  _ _ _]
+#       [_ _ _  _ _ _  _ _ _]
+
+
+class Sudoku
+
+  #
+  # @param board [Array]
+  #
+  def initialize(board)
+    @board = board
+  end
+
+  #
+  # @return [Boolean] true if board is valid
+  #
+  def solved?
```

Comment:
when writing production code, one of the main goals to focus on is "how easy is this for another engineer to understand?". it's sometimes a little fuzzy how to do this, but there are a number of best practices / benchmarks to follow

one way of making code easy to understand is breaking a large method out into smaller helper methods. in ruby, the community strives to make the code read as close to english as possible - meaning, the helper methods should verify explicitly describe what it does. for this method, we can break out the check into a method, row checks, column checks, box checks all into separate methods

side note - there are various measures of complexity that developers can help benchmark there code with. [cyclomatic complexity](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Metrics/CyclomaticComplexity), [perceived complexity](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Metrics/PerceivedComplexity), [abc method size](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Metrics/AbcSize), etc

generally, we shoot for small classes and methods that serve a single purpose

Reply:
for example, here is an example from the `Program::Earn` class in the thanx codebase:
```
  def issue_to(user)
    @user = user
    user.redis_lock do
      return nil unless issuable?
      create_reward
      send_email
      send_notification
    end
  end

  def ensure
    retire_rewards
    return unless program.active?
    return unless program.earnable?
    finalize_target
    target.call
    issue!
    finalize_program
  end
```